src/algorithms/calorimetry/ImagingTopoCluster.h

```diff
@@ -198,35 +239,56 @@ namespace eicrecon {
     }
 
     // grouping function with Breadth-First Search
-    void bfs_group(const edm4eic::CalorimeterHitCollection &hits, std::set<std::size_t> &group, std::size_t idx, std::vector<bool> &visits) const {
-      visits[idx] = true;
+    // note: template to allow Compare only known in local scope of caller
+    template<typename Compare>
+    void bfs_group(const edm4eic::CalorimeterHitCollection &hits, std::set<std::size_t,Compare>& indices, std::list<std::size_t> &group, const std::size_t idx) const {
 
       // not a qualified hit to participate clustering, stop here
       if (hits[idx].getEnergy() < m_cfg.minClusterHitEdep) {
         return;
       }
 
-      group.insert(idx);
-      size_t prev_size = 0;
+      // loop over group as it grows, until the end is stable and we reach it
+      for (auto idx1 = group.begin(); idx1 != group.end(); ++idx1) {
+        // check neighbours (note comments on loop over set above)
+        for (auto idx2 = indices.begin(); idx2 != indices.end();
+            indices.empty() ? idx2 = indices.end() : idx2) {
 
-      while (prev_size != group.size()) {
-        prev_size = group.size();
-        for (std::size_t idx1 : group) {
-          // check neighbours
-          for (std::size_t idx2 = 0; idx2 < hits.size(); ++idx2) {
-            // not a qualified hit to participate clustering, skip
-            if (hits[idx2].getEnergy() < m_cfg.minClusterHitEdep) {
-              continue;
-            }
-            if ((!visits[idx2])
-                && is_neighbour(hits[idx1], hits[idx2])) {
-              group.insert(idx2);
-              visits[idx2] = true;
-            }
+          // skip idx1 and original idx
+          // (we cannot erase idx since it would invalidate iterator in calling scope)
+          if (*idx2 == *idx1 || *idx2 == idx) {
+            idx2++;
+            continue;
+          }
+
+          // skip rest of list of hits when we're past relevant layers
+          //if (hits[*idx2].getLayer() - hits[*idx1].getLayer() > m_cfg.neighbourLayersRange) {
+          //  break;
+          //}
+
+          // not energetic enough for cluster center or other cluster hit
+          // whereas caller has removed earlier low energy hits, this removes ones that caller hasn't gotten to yet
+          if (hits[*idx2].getEnergy() < std::min(m_cfg.minClusterHitEdep,m_cfg.minClusterCenterEdep)) {
+            idx2 = indices.erase(idx2);
+            continue;
+          }
+
+          // not energetic enough for cluster hit in this group
+          if (hits[*idx2].getEnergy() < m_cfg.minClusterHitEdep) {
+            idx2++;
+            continue;
+          }
```

Comment:
Out of curiosity: would the equivalent conditions between lines 164 - 167 guarantee that the hits being processed at this point already satisfy these conditions?

Reply:
There's a list of all hits. In lines 164 - 167 the loop hasn't gotten to the hits that are being considered here. I clearly need to make my comment `whereas caller has removed earlier low energy hits, this removes ones that caller hasn't gotten to yet` clearer... Any suggestions for how to improve it?

Reply:
I agree with you that this is duplication and could be addressed.

Reply:
I think it would be good to have a clear separation of responsibilities between bfs_group and the caller without too many pre-conditions that need to be assumed. I would suggest removing the first instance of the check in the caller, as it won't really lead to extraneous execution time (bfs_group will immediately clear them right here).

Reply:
Second comment: The way it is written here the second if statement will never evaluate to true so can be removed. In the strange case that `minClusterHitEdep > minCLusterCenterEdep` all hits will just be eaten (do we have a check for this, as that should at least print a warning on init).

Reply:
> There's a set of all hits. In lines 164 - 167 the loop hasn't gotten to the hits that are being considered here. I clearly need to make my comment `whereas caller has removed earlier low energy hits, this removes ones that caller hasn't gotten to yet` clearer... Any suggestions for how to improve it?

I see! That makes sense! I think noting that we need this 2nd check since we're processing the set of all hits, so maybe some thing like `since looping over all hits, this removes low energy hits not removed earlier by caller`

Reply:
26790cf0c

Reply:
3941b5765